flytekit/common/mixins/registerable.py

```diff
@@ -171,3 +141,52 @@ def some_task()
 
         _logging.error("Could not auto-assign name")
         raise _system_exceptions.FlyteSystemException("Error looking for object while auto-assigning name.")
+
+
+class RegisterableEntity(TrackableEntity):
+    def __init__(self, *args, **kwargs):
+        self._has_registered = False
+        super(RegisterableEntity, self).__init__(*args, **kwargs)
+
+    @_abc.abstractmethod
+    def register(self, project, domain, name, version):
+        """
+        :param Text project: The project in which to register this task.
+        :param Text domain: The domain in which to register this task.
+        :param Text name: The name to give this task.
+        :param Text version: The version in which to register this task.
+        """
+        pass
+
+    @_abc.abstractmethod
+    def serialize(self):
+        """
+        Registerable entities also are required to be serialized. This allows flytekit to separate serialization from
+        the network call to Admin (mostly at least, if a Launch Plan is fetched for instance as part of another
+        workflow, it will still hit Admin).
+        """
+        pass
+
+    @property
+    def has_registered(self) -> bool:
+        return self._has_registered
+
+
+class HasDependencies(object):
```

Comment:
we are just modeling a graph - albeit poorly

Reply:
yeah.  dependency tracking in flytekit is confusing.  We have nodes, which have upstream nodes, but we also have tasks/workflows/launchplans, which also keep track of upstream entities.

I'd like to get to the point where we only use nodes to keep track of registration order and such.  This should make the topological sort easier to reason about as well.

It's not quite just the workflow graph though... if you have a workflow that statically references another workflow, somehow you'll need to know to process the second workflow first.  same for default launch plans.  we'll probably have to build up a registration graph which contains these inter-entity dependencies.  This is effectively what the entity level upstream entity array keeps track of now, but i feel like this is cleaner.  anyways, not this pr.